core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -418,102 +427,113 @@ public HashDistributionTaskSource(
+                            partitionSourceHandles.put(planNodeId, handle);
                         }
 
-                        Multimap<PlanNodeId, Split> partitionSplits = partitionToSplitsMap.computeIfAbsent(partition, (p) -> ArrayListMultimap.create());
-                        partitionSplits.put(entry.getKey(), split);
-                    }
+                        int taskPartitionId = 0;
+                        ImmutableList.Builder<TaskDescriptor> partitionTasks = ImmutableList.builder();
+                        for (Integer partition : union(partitionToSplitsMap.keySet(), partitionToExchangeSourceHandlesMap.keySet())) {
+                            ListMultimap<PlanNodeId, Split> splits = partitionToSplitsMap.getOrDefault(partition, ImmutableListMultimap.of());
+                            ListMultimap<PlanNodeId, ExchangeSourceHandle> exchangeSourceHandles = ImmutableListMultimap.<PlanNodeId, ExchangeSourceHandle>builder()
+                                    .putAll(partitionToExchangeSourceHandlesMap.getOrDefault(partition, ImmutableMultimap.of()))
+                                    // replicated exchange source will be added in postprocessTasks below
+                                    .build();
+                            Set<HostAddress> hostRequirement = partitionToNodeMap.get(partition);
+                            partitionTasks.add(new TaskDescriptor(taskPartitionId++, splits, exchangeSourceHandles, new NodeRequirements(catalogRequirement, hostRequirement, taskMemory)));
+                        }
 
-                    if (splitBatch.isLastBatch()) {
-                        splitSource.close();
-                        break;
-                    }
-                }
-            }
+                        List<TaskDescriptor> result = postprocessTasks(partitionTasks.build());
+                        finished = true;
+                        return result;
+                    },
+                    directExecutor());
+        }
 
-            Map<Integer, Multimap<PlanNodeId, ExchangeSourceHandle>> partitionToExchangeSourceHandlesMap = new HashMap<>();
-            for (Map.Entry<PlanNodeId, ExchangeSourceHandle> entry : partitionedExchangeSourceHandles.entries()) {
-                PlanNodeId planNodeId = entry.getKey();
-                ExchangeSourceHandle handle = entry.getValue();
-                int partition = handle.getPartitionId();
-                Multimap<PlanNodeId, ExchangeSourceHandle> partitionSourceHandles = partitionToExchangeSourceHandlesMap.computeIfAbsent(partition, (p) -> ArrayListMultimap.create());
-                partitionSourceHandles.put(planNodeId, handle);
-            }
+        CompletableFuture<LoadedSplits> loadSplits(PlanNodeId planNodeId, SplitSource splitSource, List<Split> loadedSplits)
+        {
+            CompletableFuture<SplitBatch> splitBatchFuture = toCompletableFuture(splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), splitBatchSize));
```

Comment:
Please avoid `CompletableFuture`. It has issues with propagating cancellation (it doesn't propagate it)